File: binding/src/sql/libqsqldatabase.cpp
#include <QSqlDatabase>
#include <QSqlDriver>
#include <QSqlDriverCreatorBase>
#include <QSqlError>
#include <QSqlIndex>
#include <QSqlQuery>
#include <QSqlRecord>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QThread>
#include <qsqldatabase.h>
#include "libqsqldatabase.h"
#include "libqsqldatabase.hxx"

QSqlDriverCreatorBase* QSqlDriverCreatorBase_new() {
	 return new VirtualQSqlDriverCreatorBase();
}

QSqlDriverCreatorBase* QSqlDriverCreatorBase_new2(const QSqlDriverCreatorBase* param1) {
	 return new VirtualQSqlDriverCreatorBase(*param1);
}

QSqlDriver* QSqlDriverCreatorBase_CreateObject(const QSqlDriverCreatorBase* self) {
	return self->createObject();
}

void QSqlDriverCreatorBase_OperatorAssign(QSqlDriverCreatorBase* self, const QSqlDriverCreatorBase* param1) {
	self->operator=(*param1);
}

// Base class handler implementation
QSqlDriver* QSqlDriverCreatorBase_QBaseCreateObject(const QSqlDriverCreatorBase* self) {
	auto* vqsqldrivercreatorbase = dynamic_cast<const VirtualQSqlDriverCreatorBase*>(self);
	if (vqsqldrivercreatorbase && vqsqldrivercreatorbase->isVirtualQSqlDriverCreatorBase) {
vqsqldrivercreatorbase->setQSqlDriverCreatorBase_CreateObject_IsBase(true);
	return vqsqldrivercreatorbase->createObject();
}
}

// Auxiliary method to allow providing re-implementation
void QSqlDriverCreatorBase_OnCreateObject(const QSqlDriverCreatorBase* self, intptr_t slot) {
	auto* vqsqldrivercreatorbase = dynamic_cast<const VirtualQSqlDriverCreatorBase*>(self);
	if (vqsqldrivercreatorbase && vqsqldrivercreatorbase->isVirtualQSqlDriverCreatorBase) {
vqsqldrivercreatorbase->setQSqlDriverCreatorBase_CreateObject_Callback(reinterpret_cast<VirtualQSqlDriverCreatorBase::QSqlDriverCreatorBase_CreateObject_Callback>(slot));
}
}

void QSqlDriverCreatorBase_Delete(QSqlDriverCreatorBase* self) {
    delete self;
}

QSqlDatabase* QSqlDatabase_new() {
	 return new QSqlDatabase();
}

QSqlDatabase* QSqlDatabase_new2(const QSqlDatabase* other) {
	 return new QSqlDatabase(*other);
}

void QSqlDatabase_OperatorAssign(QSqlDatabase* self, const QSqlDatabase* other) {
	self->operator=(*other);
}

bool QSqlDatabase_Open(QSqlDatabase* self) {
	return self->open();
}

bool QSqlDatabase_Open2(QSqlDatabase* self, const libqt_string user, const libqt_string password) {
	return self->open(QString::fromUtf8(user.data, user.len), QString::fromUtf8(password.data, password.len));
}

void QSqlDatabase_Close(QSqlDatabase* self) {
	self->close();
}

bool QSqlDatabase_IsOpen(const QSqlDatabase* self) {
	return self->isOpen();
}

bool QSqlDatabase_IsOpenError(const QSqlDatabase* self) {
	return self->isOpenError();
}

libqt_list QSqlDatabase_Tables(const QSqlDatabase* self) {
	auto _ret = self->tables();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

QSqlIndex* QSqlDatabase_PrimaryIndex(const QSqlDatabase* self, const libqt_string tablename) {
	return new QSqlIndex(self->primaryIndex(QString::fromUtf8(tablename.data, tablename.len)));
}

QSqlRecord* QSqlDatabase_Record(const QSqlDatabase* self, const libqt_string tablename) {
	return new QSqlRecord(self->record(QString::fromUtf8(tablename.data, tablename.len)));
}

QSqlQuery* QSqlDatabase_Exec(const QSqlDatabase* self) {
	return new QSqlQuery(self->exec());
}

QSqlError* QSqlDatabase_LastError(const QSqlDatabase* self) {
	return new QSqlError(self->lastError());
}

bool QSqlDatabase_IsValid(const QSqlDatabase* self) {
	return self->isValid();
}

bool QSqlDatabase_Transaction(QSqlDatabase* self) {
	return self->transaction();
}

bool QSqlDatabase_Commit(QSqlDatabase* self) {
	return self->commit();
}

bool QSqlDatabase_Rollback(QSqlDatabase* self) {
	return self->rollback();
}

void QSqlDatabase_SetDatabaseName(QSqlDatabase* self, const libqt_string name) {
	self->setDatabaseName(QString::fromUtf8(name.data, name.len));
}

void QSqlDatabase_SetUserName(QSqlDatabase* self, const libqt_string name) {
	self->setUserName(QString::fromUtf8(name.data, name.len));
}

void QSqlDatabase_SetPassword(QSqlDatabase* self, const libqt_string password) {
	self->setPassword(QString::fromUtf8(password.data, password.len));
}

void QSqlDatabase_SetHostName(QSqlDatabase* self, const libqt_string host) {
	self->setHostName(QString::fromUtf8(host.data, host.len));
}

void QSqlDatabase_SetPort(QSqlDatabase* self, int p) {
	self->setPort(p);
}

void QSqlDatabase_SetConnectOptions(QSqlDatabase* self) {
	self->setConnectOptions();
}

libqt_string QSqlDatabase_DatabaseName(const QSqlDatabase* self) {
	QString _ret = self->databaseName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDatabase_UserName(const QSqlDatabase* self) {
	QString _ret = self->userName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDatabase_Password(const QSqlDatabase* self) {
	QString _ret = self->password();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDatabase_HostName(const QSqlDatabase* self) {
	QString _ret = self->hostName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDatabase_DriverName(const QSqlDatabase* self) {
	QString _ret = self->driverName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QSqlDatabase_Port(const QSqlDatabase* self) {
	return self->port();
}

libqt_string QSqlDatabase_ConnectOptions(const QSqlDatabase* self) {
	QString _ret = self->connectOptions();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSqlDatabase_ConnectionName(const QSqlDatabase* self) {
	QString _ret = self->connectionName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSqlDatabase_SetNumericalPrecisionPolicy(QSqlDatabase* self, int precisionPolicy) {
	self->setNumericalPrecisionPolicy(static_cast<QSql::NumericalPrecisionPolicy>(precisionPolicy));
}

int QSqlDatabase_NumericalPrecisionPolicy(const QSqlDatabase* self) {
	return self->numericalPrecisionPolicy();
}

bool QSqlDatabase_MoveToThread(QSqlDatabase* self, QThread* targetThread) {
	return self->moveToThread(targetThread);
}

QThread* QSqlDatabase_Thread(const QSqlDatabase* self) {
	return self->thread();
}

QSqlDriver* QSqlDatabase_Driver(const QSqlDatabase* self) {
	return self->driver();
}

QSqlDatabase* QSqlDatabase_AddDatabase(const libqt_string typeVal) {
	return new QSqlDatabase(QSqlDatabase::addDatabase(QString::fromUtf8(typeVal.data, typeVal.len)));
}

QSqlDatabase* QSqlDatabase_AddDatabase2(QSqlDriver* driver) {
	return new QSqlDatabase(QSqlDatabase::addDatabase(driver));
}

QSqlDatabase* QSqlDatabase_CloneDatabase(const QSqlDatabase* other, const libqt_string connectionName) {
	return new QSqlDatabase(QSqlDatabase::cloneDatabase(*other, QString::fromUtf8(connectionName.data, connectionName.len)));
}

QSqlDatabase* QSqlDatabase_CloneDatabase2(const libqt_string other, const libqt_string connectionName) {
	return new QSqlDatabase(QSqlDatabase::cloneDatabase(QString::fromUtf8(other.data, other.len), QString::fromUtf8(connectionName.data, connectionName.len)));
}

QSqlDatabase* QSqlDatabase_Database() {
	return new QSqlDatabase(QSqlDatabase::database());
}

void QSqlDatabase_RemoveDatabase(const libqt_string connectionName) {
	QSqlDatabase::removeDatabase(QString::fromUtf8(connectionName.data, connectionName.len));
}

bool QSqlDatabase_Contains() {
	return QSqlDatabase::contains();
}

libqt_list QSqlDatabase_Drivers() {
	auto _ret = QSqlDatabase::drivers();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

libqt_list QSqlDatabase_ConnectionNames() {
	auto _ret = QSqlDatabase::connectionNames();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QSqlDatabase_RegisterSqlDriver(const libqt_string name, QSqlDriverCreatorBase* creator) {
	QSqlDatabase::registerSqlDriver(QString::fromUtf8(name.data, name.len), creator);
}

bool QSqlDatabase_IsDriverAvailable(const libqt_string name) {
	return QSqlDatabase::isDriverAvailable(QString::fromUtf8(name.data, name.len));
}

libqt_list QSqlDatabase_Tables1(const QSqlDatabase* self, int typeVal) {
	auto _ret = self->tables(static_cast<QSql::TableType>(typeVal));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

QSqlQuery* QSqlDatabase_Exec1(const QSqlDatabase* self, const libqt_string query) {
	return new QSqlQuery(self->exec(QString::fromUtf8(query.data, query.len)));
}

void QSqlDatabase_SetConnectOptions1(QSqlDatabase* self, const libqt_string options) {
	self->setConnectOptions(QString::fromUtf8(options.data, options.len));
}

QSqlDatabase* QSqlDatabase_AddDatabase22(const libqt_string typeVal, const libqt_string connectionName) {
	return new QSqlDatabase(QSqlDatabase::addDatabase(QString::fromUtf8(typeVal.data, typeVal.len), QString::fromUtf8(connectionName.data, connectionName.len)));
}

QSqlDatabase* QSqlDatabase_AddDatabase23(QSqlDriver* driver, const libqt_string connectionName) {
	return new QSqlDatabase(QSqlDatabase::addDatabase(driver, QString::fromUtf8(connectionName.data, connectionName.len)));
}

QSqlDatabase* QSqlDatabase_Database1(const libqt_string connectionName) {
	return new QSqlDatabase(QSqlDatabase::database(QString::fromUtf8(connectionName.data, connectionName.len)));
}

QSqlDatabase* QSqlDatabase_Database2(const libqt_string connectionName, bool open) {
	return new QSqlDatabase(QSqlDatabase::database(QString::fromUtf8(connectionName.data, connectionName.len), open));
}

bool QSqlDatabase_Contains1(const libqt_string connectionName) {
	return QSqlDatabase::contains(QString::fromUtf8(connectionName.data, connectionName.len));
}

void QSqlDatabase_Delete(QSqlDatabase* self) {
    delete self;
}

File: binding/include/sql/libqsqldatabase.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsqldrivercreatorbase.html
pub const qsqldrivercreatorbase = struct {

    /// New constructs a new QSqlDriverCreatorBase object.
    pub fn New() QtC.QSqlDriverCreatorBase {
        return qtc.QSqlDriverCreatorBase_new();
    }


    /// New2 constructs a new QSqlDriverCreatorBase object.
    pub fn New2(param1: ?*anyopaque) QtC.QSqlDriverCreatorBase {
        return qtc.QSqlDriverCreatorBase_new2(@ptrCast(param1));
    }


    pub fn CreateObject(self: ?*anyopaque, ) QtC.QSqlDriver {
        return qtc.QSqlDriverCreatorBase_CreateObject(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnCreateObject(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSqlDriver) void {
        qtc.QSqlDriverCreatorBase_OnCreateObject(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreateObject(self: ?*anyopaque, ) QtC.QSqlDriver {
        return qtc.QSqlDriverCreatorBase_QBaseCreateObject(@ptrCast(self));
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QSqlDriverCreatorBase_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSqlDriverCreatorBase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qsqldatabase.html
pub const qsqldatabase = struct {

    /// New constructs a new QSqlDatabase object.
    pub fn New() QtC.QSqlDatabase {
        return qtc.QSqlDatabase_new();
    }


    /// New2 constructs a new QSqlDatabase object.
    pub fn New2(other: ?*anyopaque) QtC.QSqlDatabase {
        return qtc.QSqlDatabase_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSqlDatabase_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Open(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_Open(@ptrCast(self));
    }

    pub fn Open2(self: ?*anyopaque, user: []const u8, password: []const u8) bool {
        const user_str = qtc.libqt_string{
    .len = user.len,
    .data = user.ptr,
};
const password_str = qtc.libqt_string{
    .len = password.len,
    .data = password.ptr,
};
return qtc.QSqlDatabase_Open2(@ptrCast(self), user_str, password_str);
    }

    pub fn Close(self: ?*anyopaque, ) void {
        qtc.QSqlDatabase_Close(@ptrCast(self));
    }

    pub fn IsOpen(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_IsOpen(@ptrCast(self));
    }

    pub fn IsOpenError(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_IsOpenError(@ptrCast(self));
    }

    pub fn Tables(self: ?*anyopaque, ) []const u8 {
        return qtc.QSqlDatabase_Tables(@ptrCast(self));
    }

    pub fn PrimaryIndex(self: ?*anyopaque, tablename: []const u8) QtC.QSqlIndex {
        const tablename_str = qtc.libqt_string{
    .len = tablename.len,
    .data = tablename.ptr,
};
return qtc.QSqlDatabase_PrimaryIndex(@ptrCast(self), tablename_str);
    }

    pub fn Record(self: ?*anyopaque, tablename: []const u8) QtC.QSqlRecord {
        const tablename_str = qtc.libqt_string{
    .len = tablename.len,
    .data = tablename.ptr,
};
return qtc.QSqlDatabase_Record(@ptrCast(self), tablename_str);
    }

    pub fn Exec(self: ?*anyopaque, ) QtC.QSqlQuery {
        return qtc.QSqlDatabase_Exec(@ptrCast(self));
    }

    pub fn LastError(self: ?*anyopaque, ) QtC.QSqlError {
        return qtc.QSqlDatabase_LastError(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_IsValid(@ptrCast(self));
    }

    pub fn Transaction(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_Transaction(@ptrCast(self));
    }

    pub fn Commit(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_Commit(@ptrCast(self));
    }

    pub fn Rollback(self: ?*anyopaque, ) bool {
        return qtc.QSqlDatabase_Rollback(@ptrCast(self));
    }

    pub fn SetDatabaseName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QSqlDatabase_SetDatabaseName(@ptrCast(self), name_str);
    }

    pub fn SetUserName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QSqlDatabase_SetUserName(@ptrCast(self), name_str);
    }

    pub fn SetPassword(self: ?*anyopaque, password: []const u8) void {
        const password_str = qtc.libqt_string{
    .len = password.len,
    .data = password.ptr,
};
qtc.QSqlDatabase_SetPassword(@ptrCast(self), password_str);
    }

    pub fn SetHostName(self: ?*anyopaque, host: []const u8) void {
        const host_str = qtc.libqt_string{
    .len = host.len,
    .data = host.ptr,
};
qtc.QSqlDatabase_SetHostName(@ptrCast(self), host_str);
    }

    pub fn SetPort(self: ?*anyopaque, p: i32) void {
        qtc.QSqlDatabase_SetPort(@ptrCast(self), p);
    }

    pub fn SetConnectOptions(self: ?*anyopaque, ) void {
        qtc.QSqlDatabase_SetConnectOptions(@ptrCast(self));
    }

    pub fn DatabaseName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_DatabaseName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.DatabaseName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn UserName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_UserName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.UserName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Password(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_Password(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.Password: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn HostName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_HostName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.HostName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn DriverName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_DriverName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.DriverName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Port(self: ?*anyopaque, ) i32 {
        return qtc.QSqlDatabase_Port(@ptrCast(self));
    }

    pub fn ConnectOptions(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_ConnectOptions(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.ConnectOptions: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ConnectionName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSqlDatabase_ConnectionName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsqldatabase.ConnectionName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetNumericalPrecisionPolicy(self: ?*anyopaque, precisionPolicy: i32) void {
        qtc.QSqlDatabase_SetNumericalPrecisionPolicy(@ptrCast(self), @intCast(precisionPolicy));
    }

    pub fn NumericalPrecisionPolicy(self: ?*anyopaque, ) i32 {
        return qtc.QSqlDatabase_NumericalPrecisionPolicy(@ptrCast(self));
    }

    pub fn MoveToThread(self: ?*anyopaque, targetThread: ?*anyopaque) bool {
        return qtc.QSqlDatabase_MoveToThread(@ptrCast(self), @ptrCast(targetThread));
    }

    pub fn Thread(self: ?*anyopaque, ) QtC.QThread {
        return qtc.QSqlDatabase_Thread(@ptrCast(self));
    }

    pub fn Driver(self: ?*anyopaque, ) QtC.QSqlDriver {
        return qtc.QSqlDatabase_Driver(@ptrCast(self));
    }

    pub fn AddDatabase(typeVal: []const u8) QtC.QSqlDatabase {
        const typeVal_str = qtc.libqt_string{
    .len = typeVal.len,
    .data = typeVal.ptr,
};
return qtc.QSqlDatabase_AddDatabase(typeVal_str);
    }

    pub fn AddDatabase2(driver: ?*anyopaque) QtC.QSqlDatabase {
        return qtc.QSqlDatabase_AddDatabase2(@ptrCast(driver));
    }

    pub fn CloneDatabase(other: ?*anyopaque, connectionName: []const u8) QtC.QSqlDatabase {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_CloneDatabase(@ptrCast(other), connectionName_str);
    }

    pub fn CloneDatabase2(other: []const u8, connectionName: []const u8) QtC.QSqlDatabase {
        const other_str = qtc.libqt_string{
    .len = other.len,
    .data = other.ptr,
};
const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_CloneDatabase2(other_str, connectionName_str);
    }

    pub fn Database() QtC.QSqlDatabase {
        return qtc.QSqlDatabase_Database();
    }

    pub fn RemoveDatabase(connectionName: []const u8) void {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
qtc.QSqlDatabase_RemoveDatabase(connectionName_str);
    }

    pub fn Contains() bool {
        return qtc.QSqlDatabase_Contains();
    }

    pub fn Drivers() []const u8 {
        return qtc.QSqlDatabase_Drivers();
    }

    pub fn ConnectionNames() []const u8 {
        return qtc.QSqlDatabase_ConnectionNames();
    }

    pub fn RegisterSqlDriver(name: []const u8, creator: ?*anyopaque) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QSqlDatabase_RegisterSqlDriver(name_str, @ptrCast(creator));
    }

    pub fn IsDriverAvailable(name: []const u8) bool {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QSqlDatabase_IsDriverAvailable(name_str);
    }

    pub fn Tables1(self: ?*anyopaque, typeVal: i32) []const u8 {
        return qtc.QSqlDatabase_Tables1(@ptrCast(self), @intCast(typeVal));
    }

    pub fn Exec1(self: ?*anyopaque, query: []const u8) QtC.QSqlQuery {
        const query_str = qtc.libqt_string{
    .len = query.len,
    .data = query.ptr,
};
return qtc.QSqlDatabase_Exec1(@ptrCast(self), query_str);
    }

    pub fn SetConnectOptions1(self: ?*anyopaque, options: []const u8) void {
        const options_str = qtc.libqt_string{
    .len = options.len,
    .data = options.ptr,
};
qtc.QSqlDatabase_SetConnectOptions1(@ptrCast(self), options_str);
    }

    pub fn AddDatabase22(typeVal: []const u8, connectionName: []const u8) QtC.QSqlDatabase {
        const typeVal_str = qtc.libqt_string{
    .len = typeVal.len,
    .data = typeVal.ptr,
};
const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_AddDatabase22(typeVal_str, connectionName_str);
    }

    pub fn AddDatabase23(driver: ?*anyopaque, connectionName: []const u8) QtC.QSqlDatabase {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_AddDatabase23(@ptrCast(driver), connectionName_str);
    }

    pub fn Database1(connectionName: []const u8) QtC.QSqlDatabase {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_Database1(connectionName_str);
    }

    pub fn Database2(connectionName: []const u8, open: bool) QtC.QSqlDatabase {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_Database2(connectionName_str, open);
    }

    pub fn Contains1(connectionName: []const u8) bool {
        const connectionName_str = qtc.libqt_string{
    .len = connectionName.len,
    .data = connectionName.ptr,
};
return qtc.QSqlDatabase_Contains1(connectionName_str);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSqlDatabase_Delete(@ptrCast(self));
    }
};
